### PR TITLE
[CHK-1863] feat(npg): handle npg refund

### PR DESCRIPTION
**Pull request by ciuffagianluca**
Handle NPG refund request

depends on https://github.com/pagopa/pagopa-ecommerce-commons/pull/125

#### List of Changes

Integrate NpgClient into dispatcher service
Handle npg refund request
Handle npg refund request response ok and ko
Add tests

#### Motivation and Context

This PR is necessary for the npg transaction workflow portion where a refund is requestes

#### How Has This Been Tested?

JUnit tests has been ran

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt [hunk 1]
```diff
@@ -220,6 +243,28 @@ fun refundTransaction(
...
+  transaction: BaseTransactionWithRequestedAuthorization,
+  refundResponse: RefundResponseDto,
+  transactionsEventStoreRepository: TransactionsEventStoreRepository<TransactionRefundedData>,
+  transactionsViewRepository: TransactionsViewRepository
+): Mono<BaseTransaction> {
+  logger.info(
+    "Transaction requestRefund for transaction ${transaction.transactionId} NPG refund status [${refundResponse.operationId!!}]")
+
+  return when (refundResponse.operationId) { // TODO Check this response value, which are possible?
+    "operationId" -> { // TODO this value need to be retrieved from transaction event data
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt:256]: This field is an NPG identifier (an UUID), why this when here?

Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt:256]: Hi @pietro-tota , nexi specs are a little bit ambiguous in this response.
We know that refundResponse should contain 2 field:

- operationId
- operationDateTime

Both aren't mandatory. So what i guess here is this business logic:
If I receive refundResponse with operationId that matches the operationId of the authorization i'm asking for refunding, i can guess that refund operation is ok. If your doubt is about constant string "operationId", don't forget that this is a placeholder for a data that we'll retrieve from authorization request data when we'll save it from the patch. This task will be faced in the next sprint, so by now I put here a constant.

Comment (infantesimone) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt:256]: Comparing the docs with the openapi specifications that we have on the other api, I think those fields are mandatory so we can avoid the check (unfortunately we do not have the openapi specifications for refund api )

File: src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/RefundServiceTests.kt [hunk 2]
```diff
@@ -1,26 +1,193 @@
...
-  @InjectMocks private lateinit var refundService: RefundService
+  @Test
+  fun requestRefund_200_npg() {
+    val operationId = "operationID"
+    val idempotenceKey = "idempotenceKey"
+    val amount = BigDecimal.valueOf(1000)
+
+    // Precondition
+    Mockito.`when`(npgClient.refundPayment(any(), any(), any(), any(), any()))
+      .thenReturn(Mono.just(getMockedNpgRefundResponse(operationId)))
```
Comment (giovanniberti) [inline src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/RefundServiceTests.kt:42]: You can use `given` instead so you don't have to escape the method name

File: pom.xml [hunk 3]
```diff
@@ -507,8 +507,8 @@
 				<configuration>
 					<skipCheckoutIfExists>false</skipCheckoutIfExists>
 					<connectionUrl>scm:git:https://github.com/pagopa/pagopa-ecommerce-commons.git</connectionUrl>
-					<scmVersionType>tag</scmVersionType>
-					<scmVersion>${pagopa-ecommerce-commons.version}</scmVersion>
+					<scmVersionType>branch</scmVersionType>
+					<scmVersion>CHK-1862-implment-refund-payment</scmVersion>
```
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply pom.xml:511]: Did it!

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt [hunk 4]
```diff
@@ -171,9 +176,21 @@ fun refundTransaction(
             Pair(refundResponse, transaction)
           }
         else ->
-          Mono.error(
-            RuntimeException(
-              "Refund error for transaction ${transaction.transactionId} - unhandled payment-gateway"))
+          when (transaction.transactionAuthorizationRequestData.paymentMethodName) {
+            NpgClient.PaymentMethod.CARDS.serviceName ->
```
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt:180]: I expose this point to @infantesimone and we agree on the fact that we have to add `NPG` variant to `TransactionAuthorizationRequestData.PaymentGateway`. Anyway we decide to merge this PR (once approved) then face that task and this refactoring in the next sprint

File: pom.xml [hunk 5]
```diff
@@ -19,7 +19,7 @@
 		<kotlinx-coroutines.version>1.6.2</kotlinx-coroutines.version>
 		<spring-cloud-azure.version>4.0.0</spring-cloud-azure.version>
 		<jacoco.version>0.8.8</jacoco.version>
-		<pagopa-ecommerce-commons.version>0.19.5</pagopa-ecommerce-commons.version>
+		<pagopa-ecommerce-commons.version>0.22.1</pagopa-ecommerce-commons.version>
```
Comment (Fernando-Granato) [inline pom.xml:22]: Modified this version with new version 0.23.0

**Review by Fernando-Granato — COMMENTED** [review]
I left a comment

File: src/main/resources/application.properties [hunk 6]
```diff
@@ -33,6 +33,10 @@ paymentTransactionsGateway.uri=${PAYMENT_TRANSACTIONS_GATEWAY_URI}/
 paymentTransactionsGateway.readTimeout=${PAYMENT_TRANSACTIONS_GATEWAY_READ_TIMEOUT}
 paymentTransactionsGateway.connectionTimeout=${PAYMENT_TRANSACTIONS_GATEWAY_CONNECTION_TIMEOUT}
 paymentTransactionsGateway.apiKey=${PAYMENT_TRANSACTIONS_GATEWAY_API_KEY}
+npg.uri=${NPG_URI}/
```
Comment (pietro-tota) [inline src/main/resources/application.properties:36]: why this trailing `/`

Comment (ciuffagianluca) [reply src/main/resources/application.properties:36]: It's a typo. To be sure that it works, i'll redeploy on dev env after fix

Comment (ciuffagianluca) [reply src/main/resources/application.properties]: It works without "/"

**Comment by pietro-tota** [comment]
Can we close this PR since it's superseeded by https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/154?

**Comment by ciuffagianluca** [comment]
Dismissed in favour of #154

**Pull request closed without merging**